### PR TITLE
Peer health checks

### PR DESCRIPTION
**Pull request by shubham3121**
## Description
Please include a summary of the change, the motivation, and any additional context that will help others understand your PR. If it closes one or more open issues, [please tag them as described here](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

## Affected Dependencies
List any dependencies that are required for this change.

- track ping status, ping message and ping timestamp at NodePeer level
- add a utility track peer route healthcheck
- add a network service to find peer
- add an integration test for peers health check

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

**Comment by khoaguin** [comment]
TODO in this PR:
- Integrate `background_tasks` in orchestra

2. for `run_peer_health_checks` background task
- What is the ping status of a peer if association request is submitted but not accepted?
- What is the ping status if the association request is accepted?
- If one peer node goes down, what is the ping status (may not be checked with a Python Node. Try starting a server with HTTP connection)
- what is the ping status when a peer is deleted -> create remote client on that node peer -> check peer list -> peer not found

3. Since we can get domain health status now with the new peers health check functionalities, we can just grab the online domains from the database and no longer need to check it like currently in `DomainRegistry.online_domains` using a `ThreadPoolExecutor`

File: packages/syft/src/syft/service/network/network_service.py [hunk 1]
```diff
@@ -237,19 +237,39 @@ def add_peer(
...
+        )
+        if (
+            association_requests
+            and association_requests[-1].status == RequestStatus.PENDING
+        ):
+            return SyftError(
+                message="There is already a pending association request for this peer"
+            )
+        # only create and submit a new request if there is no requests yet
+        # or all previous requests have been rejected
```
Comment (khoaguin) [inline packages/syft/src/syft/service/network/network_service.py]: @kiendang please review this

Comment (khoaguin) [reply packages/syft/src/syft/service/network/network_service.py]: This is the fix for the association requests that we discussed earlier
- Check the last association request by timestamp (`updated_at`) and assign `PEER_ASSOCIATION_PENDING` only when the last request is pending
- Do not allow new association requests if there is a pending request from a peer / peer already associated
- Only create new requests if the last request from the peer has been denied or the peer does not submit any request yet

Comment (kiendang) [reply packages/syft/src/syft/service/network/network_service.py]: Looks good to me apart from a nitpick change. Needs a second pair of eyes from @shubham3121 too.

File: packages/syft/src/syft/service/network/network_service.py [hunk 2]
```diff
@@ -228,19 +237,39 @@ def add_peer(
                 message="verify_key does not match the remote node's verify_key for add_peer"
             )
 
+        # check if the peer already is a node peer
+        existed_peer = self.stash.get_by_uid(context.node.verify_key, peer.id)
+        if existed_peer.is_err():
+            return SyftError(
+                message=f"Failed to query peer from stash: {existed_peer.err()}"
+            )
```
Comment (kiendang) [inline packages/syft/src/syft/service/network/network_service.py]: We use python>=3.10 now so can safely do this go-like thing

```suggestion
        if (err := existed_peer.is_err()):
            return SyftError(
                message=f"Failed to query peer from stash: {err}"
            )
```

but I'm not sure about including the error/exception here. @shubham3121?

File: packages/hagrid/hagrid/orchestra.py [hunk 3]
```diff
@@ -272,6 +272,7 @@ def deploy_to_python(
         "n_consumers": n_consumers,
         "create_producer": create_producer,
         "association_request_auto_approval": association_request_auto_approval,
+        "background_tasks": True,
```
Comment (shubham3121) [inline packages/hagrid/hagrid/orchestra.py]: suggestion: we need to propagate this `background_tasks` flag to orchestra.launch

File: packages/syft/src/syft/service/network/network_service.py [hunk 4]
```diff
@@ -170,6 +179,33 @@ def exchange_credentials_with(
...
+
+        remote_node_peer = NodePeer.from_client(remote_client)
+        existing_peer_result = self.stash.get_by_uid(
+            context.credentials.verify_key, remote_node_peer.id
+        )
+        if (
+            existing_peer_result.is_ok()
+            and (existing_peer := existing_peer_result.ok()) is not None
+        ):
+            msg = [f"Routes already exchanged."]
```
Comment (shubham3121) [inline packages/syft/src/syft/service/network/network_service.py]: nitpick: remove fstring `f`

File: packages/syft/src/syft/service/network/node_peer.py [hunk 5]
```diff
@@ -66,6 +75,9 @@ class NodePeer(SyftObject):
     node_routes: list[NodeRouteType] = []
     node_type: NodeType
     admin_email: str
+    ping_status: NodePeerConnectionStatus | None = None
+    ping_status_message: str | None = None
+    pinged_timestamp: DateTime | None = None
```
Comment (shubham3121) [inline packages/syft/src/syft/service/network/node_peer.py:87]: Can we add these three vars to __repr__attrs__

File: packages/syft/src/syft/service/network/utils.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                    peer.ping_status_message = f"Peer '{peer.name}''s ping status: {peer.ping_status.value.lower()}"
+
+            result = network_stash.update(
+                credentials=context.node.verify_key,
+                peer=peer,
+                has_permission=True,
+            )
+
+            if result.is_err():
+                logging.info(f"Failed to update peer in stash: {result.err()}")
```
Comment (shubham3121) [inline packages/syft/src/syft/service/network/utils.py]: can we replace use of logging with loguru

Comment (shubham3121) [reply packages/syft/src/syft/service/network/utils.py]:
```python
from loguru import logger
logger.info()
```

File: packages/syft/src/syft/service/request/request_service.py [hunk 7]
```diff
@@ -288,6 +288,19 @@ def save(
             message=f"Failed to update Request: <{request.id}>. Error: {result.err()}"
         )
 
+    @service_method(
+        path="request.delete_by_id",
+        name="delete_by_id",
+    )
+    def delete_by_id(
```
Comment (shubham3121) [inline packages/syft/src/syft/service/request/request_service.py]: nitpick: rename to `delete_by_uid`

**Comment by shubham3121** [comment]
Great work @khoaguin and @kiendang !!! Left a few minor comments.

File: packages/hagrid/hagrid/orchestra.py (hunk 3 above)
Comment (shubham3121) [reply packages/hagrid/hagrid/orchestra.py]: @khoaguin I think we missed addressing this comment. Need to integrate background_tasks to def launch and then propogate it to other deployment types.

Comment (khoaguin) [reply packages/hagrid/hagrid/orchestra.py]: Thank you very much @shubham3121
Integrating `background_tasks` flag to `orchestra.launch` in this commit https://github.com/OpenMined/PySyft/pull/8795/commits/8182a1480a6dfcf3f56568428b35aa644e475807

**Review by shubham3121 — COMMENTED** [review]
Great work @khoaguin and thanks for the help @kiendang !! 🎉 💯

**Pull request merged**